Write about Keccak and SHA-3

EdOverflow opened the pull request:
I am looking forward to your feedback.

lvh commented on Crypto101.org:
```diff
@@ -3026,6 +3026,22 @@ Somitra Kumar Sanadhya and Palash Sarkar were able to cause collisions with
+
+Although SHA-3 sounds like it might come from the same family as SHA-2, the two are designed very differently. SHA-3 is very efficient in hardware, but is relatively slow in software in comparison to SHA-2. Security wise, there is nothing to say that SHA-3 is more secure than SHA-2. It is simply a possible replacement in case SHA-2 is broken.
+
+The SHA-3 hash functions were introduced in Python version 3.6 and can be implemented as follows:
```
Prefer "used" instead of "implemented".

lvh commented on Crypto101.org:
```diff
@@ -3026,6 +3026,22 @@ Somitra Kumar Sanadhya and Palash Sarkar were able to cause collisions with
+
+Keccak is a family of sponge functions designed by Guido Bertoni, Joan Daemen, Gilles Van Assche and Michaël Peeters, which won NIST's Secure Hash Algorithm Competition in 2012. Keccak has since been standardized in form of the SHA3-224, SHA3-256, SHA3-384 and SHA3-512 hash functions.
+
+Although SHA-3 sounds like it might come from the same family as SHA-2, the two are designed very differently. SHA-3 is very efficient in hardware, but is relatively slow in software in comparison to SHA-2. Security wise, there is nothing to say that SHA-3 is more secure than SHA-2. It is simply a possible replacement in case SHA-2 is broken.
```
This makes some claims I'm not sure I'm comfortable with as-is.

Performance numbers should be cited. If security claims are made, they should be put into context. While I generally don't think length-extension attacks on Merkle-Damgard hashes are a flaw in the hash function, it's important to note that a) NIST considered a drastically different design to be intrinsically desirable, and b) not being vulnerable to length-extension attacks (a fortiori: having a direct PRF mode) was considered a conditio sine qua non for SHA-3 contestants.

lvh replied:
Note that the hash functions section already talks about length extension attacks and how SHA-3 defends against them, so you don't necessarily need to repeat that here. A reference to that would be nice though.

EdOverflow replied:
> Performance numbers should be cited.

I will add references to https://bench.cr.yp.to/results-sha3.html and https://theses.lib.vt.edu/theses/available/etd-05172011-141328/unrestricted/Huang_S_T_2011.pdf.

EdOverflow commented:
Hi @lvh,

Sorry for the long wait, I have been super busy! I have added references to the corresponding papers and changed `implemented` to `used`. Please let me know if I need to make any changes.

lvh commented:
yes you should totally feel bad about latency on the unremuniated work you do in your free time ;)

lvh commented:
LGTM. Thanks!
